### ISSUE TITLE
Lesson Feedback: Mutable file system - Lesson 4 (Add a file to MFS)

### ISSUE THREAD

**Issue by NukeManDan** [ProtoSchool/protoschool.github.io#252]
Have a question or suggestion regarding a specific ProtoSchool lesson? Please use this template to share it!

**URL of the lesson that's confusing:**
eg https://proto.school/#/mutable-file-system/04

**What's confusing about this lesson?**
Malformed solution leaves the "submit" button hanging.

```javascript
const run = async (files) => {
  for (let file of files) {
    await ipfs.files.write('/'+file, file, { create: true })  }
}

return run
```
Note the file name is malformed : '/'+file 

**What additional context could we provide to help you succeed?**

An "incorrect" prompt vs a loading loop.

Here is my console: 
![image](https://user-images.githubusercontent.com/35669742/60750438-9d3c5180-9f65-11e9-8b46-10d99db1b08d.png)

**Comment by terichadbourne** [comment]
Thanks for documenting this @NukeManDan! 

_Apologies for what has turned out to be a very long braindump as I experiment..._ 

I initially read your comment as though the malformedness was about not putting spaces before and after the `+`, but this works: 

```js
const run = async (files) => {
  for (let file of files) {
    await ipfs.files.write('/'+file.name, file, { create: true })
  }
}

return run
```
so I can tell that the error is using just `file` (an object) instead of `file.name` (a string). I do get the same endless loading as you when doing it like this: 
```js
    await ipfs.files.write('/'+file, file, { create: true })
```
and in my browser (Chrome) the error is: 

![image](https://user-images.githubusercontent.com/19171465/64460054-74e8e480-d0c7-11e9-9078-02638c8d67a7.png)

It's getting stuck on trying to return `result.error` so it can print the error generated by the system.  Here's the relevant code: https://github.com/ProtoSchool/protoschool.github.io/blob/code/src/tutorials/0004/04.vue#L52

I tried logging just `result` in the console and got this: 

![image](https://user-images.githubusercontent.com/19171465/64460167-bd080700-d0c7-11e9-8d3d-75f39656c664.png)

So I assume `result` doesn't ever get passed from [the `run` function](https://github.com/ProtoSchool/protoschool.github.io/blob/code/src/components/Lesson.vue#L273) in the Lesson component to [the `validate` function](https://github.com/ProtoSchool/protoschool.github.io/blob/code/src/tutorials/0004/04.vue#L17) in this specific lesson's file. 

I thought perhaps`ipfs.files.write` doesn't know what to do when you mess up and pass it an object instead of a string for the `path`, but this produced a specific error: 
![image](https://user-images.githubusercontent.com/19171465/64461216-8f708d00-d0ca-11e9-9d67-ce9382ef6099.png)

And then I realized concatenating a string and an object would make something that's not an object, so what you are actually passing in as `path` is the string `/[object Object]`.

This test also hangs like the real thing did. 
```js
    await ipfs.files.write("/[object Object]", file, { create: true })
```
Soooo maybe IPFS doesn't know what to do if you pass a string with a space in it or a string with square brackets in it to `ipfs.files.write`? 

This also hangs (string with a space but without brackets) : 
```js
    await ipfs.files.write("/blah blah", file, { create: true })
```
as does this (string with brackets but no space)
```js
    await ipfs.files.write("/blah[b]lah", file, { create: true })
```

I believe I need help from someone more familiar with IPFS and filepaths to help me hone in on the right theory here. Any volunteers? :) 

It is, of course, also possible that there's something wrong with our own ProtoSchool code 😂, but if it is indeed an IPFS error with no error message then it would be great to add an error message to IPFS if possible, which would then automagically dipslay itself here.

**Comment by terichadbourne** [comment]
I've opened an issue in the MFS repo [#61](https://github.com/ipfs/js-ipfs-mfs/issues/61) at  @achingbrain's request and he's offered to take a look next week to see whether this is indeed an IPFS problem as opposed to a ProtoSchool problem. (Thanks, Alex!) We'll let this sit until we hear back. 

@NukeManDan wins the prize for most perplexing lesson feedback issue to date. 🏆

**Comment by terichadbourne** [comment]
@achingbrain & I met today and confirmed this was a ProtoSchool issue, not an IPFS issue. 

In MFS Lesson 4 validation code we're missing a specific conditional so it's jumping to trying to return `result.error`. Normally this catch-all is meant to display errors generated by IPFS or by syntax problems after first checking for the mostly likely errors we anticipate users making, but because neither of those types of errors exist and because this coding challenge doesn't ask the user to return a value from their interior function, it's actually looking for a result of that interior function that doesn't exist. @achingbrain has a suggestion for updating our eval code in a way that would help us distinguish between the various types of results and errors our code could currently refer to. I'll let him add his analysis here in more detail when he has a minute.

**Comment by terichadbourne** [comment]
I believe this was closed when we addressed #61. Please let us know if you encounter it again!
